### PR TITLE
Fix clang reject valid C++ code after d999ce0302f06d250f6d496b56a5a5f

### PR DESCRIPTION
**Pull request by hokein**
The incremental processing mode doesn't seem to work well for C++, see the https://github.com/llvm/llvm-project/pull/89804#issuecomment-2149840711 for details.

### CODE REVIEW

**Review by vgvassilev — APPROVED** [review]
LGTM, can you include the produced errors and the steps to reproduce the failure in the commit log? Or refer to the github post describing it?

**Comment by hokein** [comment]
unfortunately, this seems to break an existing test:

```
       OK ] InterpreterTest.IncrementalInputTopLevelDecls (66 ms)
[ RUN      ] InterpreterTest.Errors
ClangReplInterpreterTests: /usr/local/google/home/hokein/workspace/llvm-project/clang/lib/Sema/IdentifierResolver.cpp:228: void clang::IdentifierResolver::RemoveDecl(NamedDecl *): Assertion `Ptr && "Didn't find this decl on its identifier's chain!"' failed.
 #0 0x00005591835285e1 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (tools/clang/unittests/Interpreter/ClangReplInterpreterTests+0xa4e45e1)
 #1 0x0000559183528b9b PrintStackTraceSignalHandler(void*) Signals.cpp:0:0
 #2 0x0000559183526356 llvm::sys::RunSignalHandlers() (tools/clang/unittests/Interpreter/ClangReplInterpreterTests+0xa4e2356)
 #3 0x0000559183529d35 SignalHandler(int) Signals.cpp:0:0
 #4 0x00007f5baf25a510 (/lib/x86_64-linux-gnu/libc.so.6+0x3c510)
 #5 0x00007f5baf2a816
```

**Comment by vgvassilev** [comment]
Oh, we need to adjust https://github.com/root-project/root/blob/be5d34934de883270683030b3af2cd1195d17ea8/cmake/modules/RootMacros.cmake#L272 to skip in case of C++...

**Comment by hokein** [comment]
> Oh, we need to adjust https://github.com/root-project/root/blob/be5d34934de883270683030b3af2cd1195d17ea8/cmake/modules/RootMacros.cmake#L272 to skip in case of C++...

The link points to an irrelevant project, I assume you mean here https://github.com/llvm/llvm-project/blob/main/clang/lib/Interpreter/IncrementalParser.cpp#L416? If we need to skip for C++, I think we should do it for ObjectiveC as well, like

```
if (ND->getDeclName().getFETokenInfo()  &&
    !(!PP.isIncrementalProcessingEnabled() || getLangOpts().ObjC || getLangOpts().CPlusPlus)) {
   ...
}
```

**Comment by vgvassilev** [comment]
> > Oh, we need to adjust https://github.com/root-project/root/blob/be5d34934de883270683030b3af2cd1195d17ea8/cmake/modules/RootMacros.cmake#L272 to skip in case of C++...
> 
> The link points to an irrelevant project, I assume you mean here https://github.com/llvm/llvm-project/blob/main/clang/lib/Interpreter/IncrementalParser.cpp#L416? If we need to skip for C++, I think we should do it for ObjectiveC as well, like
> 
> ```
> if (ND->getDeclName().getFETokenInfo()  &&
>     !(!PP.isIncrementalProcessingEnabled() || getLangOpts().ObjC || getLangOpts().CPlusPlus)) {
>    ...
> }
> ```

Not sure how that happened. I meant https://github.com/llvm/llvm-project/blob/70550cd6aa9f2587e166d6ab9636192af3f3264d/clang/lib/Interpreter/IncrementalParser.cpp#L416-L417

**Comment by hokein** [comment]
I have update a new version, please take a second look.

File: clang/lib/Interpreter/IncrementalParser.cpp [hunk 1]
```diff
@@ -413,7 +413,9 @@ void IncrementalParser::CleanUpPTU(PartialTranslationUnit &PTU) {
     if (!ND)
       continue;
     // Check if we need to clean up the IdResolver chain.
-    if (ND->getDeclName().getFETokenInfo())
+    if (ND->getDeclName().getFETokenInfo() &&
+        CI->getPreprocessor().isIncrementalProcessingEnabled() &&
```
Comment (vgvassilev) [inline clang/lib/Interpreter/IncrementalParser.cpp:417]: Here the incremental processing is guaranteed. We do not need to check.

Comment (hokein) [reply clang/lib/Interpreter/IncrementalParser.cpp]: good point. Removed.

**Comment by hokein** [comment]
I'm landing it now to unblock our integration. I'm happy to address any post comments.

**Pull request merged**